bin/generate_east_asian_width

```diff
@@ -13,66 +25,32 @@ open(ARGV.first, 'rt') do |f|
     unicode_version = nil
   end
 
-  list = []
+  widths = []
+  letter_modifiers = []
   f.each_line do |line|
-    next unless m = line.match(/^(\h+)(?:\.\.(\h+))?\s*;\s*(\w+)\s+#.+/)
+    next unless /^(?<first>\h+)(?:\.\.(?<last>\h+))?\s*;\s*(?<type>\w+)\s+# +(?<category>[^ ]+)/ =~ line
 
-    first = m[1].to_i(16)
-    last = m[2]&.to_i(16) || first
-    type = m[3].to_sym
-    if !list.empty? and (list.last[:range].last + 1) == first and list.last[:type] == type
-      list.last[:range] = (list.last[:range].first..last)
-    else
-      # [\u{D800}-\u{DFFF}] cause error.
-      unless ((0xD800..0xDFFF).to_a & (first..last).to_a).empty?
-        unless (first..0xD7FF).to_a.empty?
-          list << {
-            range: (first..0xD7FF),
-            type: type.to_sym
-          }
-        end
-        unless (0xE000..last).to_a.empty?
-          list << {
-            range: (first..0xD7FF),
-            type: type.to_sym
-          }
-        end
-      else
-        list << {
-          range: (first..last),
-          type: type.to_sym
-        }
-      end
-    end
+    range = first.to_i(16)..(last || first).to_i(16)
+    widths.fill(unicode_width(type, category), range)
   end
-  grouped = list.group_by { |item| item[:type] }.map { |item| [item.first, item.last.map { |row| row[:range] }] }.to_h
-  grouped = %i{F H W Na A N}.map { |type| [type, grouped[type]] }
-  puts <<EOH
-class Reline::Unicode::EastAsianWidth
-  # This is based on EastAsianWidth.txt
-  # UNICODE_VERSION = #{unicode_version ? "'#{unicode_version}'" : 'nil'}
 
-EOH
-  puts grouped.map { |item|
-    type, ranges = item
-    output =  "  # %s\n" %
-      case type
-      when :F  then 'Fullwidth'
-      when :H  then 'Halfwidth'
-      when :W  then 'Wide'
-      when :Na then 'Narrow'
-      when :A  then 'Ambiguous'
-      when :N  then 'Neutral'
-      end
-    output += "  TYPE_%s = /^[\#{ %%W(\n" % type.upcase
-    output += ranges.map { |range|
-      if range.first == range.last
-        '    \u{%04X}' % range.first
-      else
-        '    \u{%04X}-\u{%04X}' % [range.first, range.last]
-      end
-    }.join("\n")
-    output += "\n  ).join }]/\n"
-  }.join("\n")
-  puts 'end'
+  # EscapedPairs
+  [*0x00..0x1F, 0x7F].each { |ord| widths[ord] = 2 }
+  # printable ASCII chars
+  (0x20..0x7E).each { |ord| widths[ord] = 1 }
+
+  chunks = widths.each_with_index.chunk { |width, _idx| width || 1 }
+  chunk_last_ords = chunks.map { |width, chunk| [chunk.last.last, width] }
+  chunk_last_ords << [0x7fffffff, 1]
```

Comment:
I think `chunk_last_ords` should be up to `0x10FFFFF`.

> In the Unicode Standard, the codespace consists of the integers from 0 to 10FFFF
2.4 Code Points and Characters
https://www.unicode.org/versions/Unicode15.1.0/ch02.pdf

Reply:
I added this value(max value of int32) as an alternative of infinity. This way, we don't need to test `0x10ffff` not to raise `no method error`.
I think it is not so obvious that `chunk_index = EastAsianWidth::CHUNK_LAST.bsearch_index { |o| ord <= o }` does not return nil when `CHUNK_LAST.last == 0x10FFFF` and `ord == 0x10FFFF `.
Of course we can change it to `0x10FFFF` and add a boundary value test. What do you think?

Reply:
Ah, I see. I agree that using `0x7fffffff` is good, just as you suggested.